Source/SlimDetours/Thread.c

```diff
@@ -6,107 +6,123 @@
-    }
+        ThreadHandle = hNextThread;
+        ClosePrevThread = TRUE;
+
+        THREAD_BASIC_INFORMATION BasicInformation;
+        if (!NT_SUCCESS(NtQueryInformationThread(
+            ThreadHandle,
+            ThreadBasicInformation,
+            &BasicInformation,
+            sizeof(BasicInformation),
+            NULL
+        )))
+        {
+            continue;
+        }
 
-    /* Create handle array */
-    Buffer = (PHANDLE)detour_memory_alloc(ThreadCount * sizeof(HANDLE));
-    if (Buffer == NULL)
-    {
-        Status = STATUS_NO_MEMORY;
-        goto _Exit;
-    }
+        /* Skip the current thread */
+        if (BasicInformation.ClientId.UniqueThread == CurrentTID)
+        {
+            continue;
+        }
 
-    /* Suspend threads */
-    SuspendedCount = 0;
-    CurrentTID = (HANDLE)(ULONG_PTR)NtCurrentThreadId();
-    for (i = 0; i < pCurrentSPI->NumberOfThreads; i++)
-    {
-        if (pSTI[i].ClientId.UniqueThread == CurrentTID ||
-            !NT_SUCCESS(NtOpenThread(&ThreadHandle,
-                                     THREAD_SUSPEND_RESUME | THREAD_GET_CONTEXT | THREAD_SET_CONTEXT,
-                                     &ObjectAttributes,
-                                     &pSTI[i].ClientId)))
+        if (!NT_SUCCESS(NtSuspendThread(ThreadHandle, NULL)))
         {
             continue;
         }
-        if (NT_SUCCESS(NtSuspendThread(ThreadHandle, NULL)))
+
+        ClosePrevThread = FALSE;
+
+        Status = STATUS_SUCCESS;
+        if (Buffer == NULL)
         {
-            _Analysis_assume_(SuspendedCount < ThreadCount);
-            Buffer[SuspendedCount++] = ThreadHandle;
-        } else
+            BufferCapacity = INITIAL_THREAD_CAPACITY;
+            Buffer = (PHANDLE)detour_memory_alloc(BufferCapacity * sizeof(HANDLE));
+            if (Buffer == NULL)
+            {
+                Status = STATUS_NO_MEMORY;
+            }
+        } else if (SuspendedCount >= BufferCapacity)
         {
+            BufferCapacity *= 2;
+            LPHANDLE p = (PHANDLE)detour_memory_realloc(Buffer, BufferCapacity * sizeof(HANDLE));
+            if (p)
+            {
+                Buffer = p;
+            }
+            else
+            {
+                Status = STATUS_NO_MEMORY;
+            }
+        }
+
+        if (!NT_SUCCESS(Status))
+        {
+            NtResumeThread(ThreadHandle, NULL);
             NtClose(ThreadHandle);
+            break;
         }
+
+        // Perform a synchronous operation to make sure the thread really is suspended.
+        // https://devblogs.microsoft.com/oldnewthing/20150205-00/?p=44743
+        CONTEXT cxt;
+        cxt.ContextFlags = CONTEXT_CONTROL;
+        NtGetContextThread(ThreadHandle, &cxt);
+
+        Buffer[SuspendedCount++] = ThreadHandle;
     }
 
-    /* Return suspended thread handles */
-    if (SuspendedCount == 0)
+    if (!NT_SUCCESS(Status) && Buffer != NULL)
     {
+        for (UINT i = 0; i < SuspendedCount; ++i)

```

Comment:
```suggestion
        UINT i;
        for (i = 0; i < SuspendedCount; ++i)
```
This is C source, so declare `i` outside `for` statement is more compliant with C standard.

Reply:
Thanks for the review, I'll take a closer look and will incorporate the changes soon.

Regarding this comment and moving variables to the beginning of the function: declaring a variable in a for loop is [valid in C99](https://en.cppreference.com/w/c/language/for), and the codebase already makes use of it, for example:
https://github.com/KNSoft/KNSoft.SlimDetours/blob/4cc236ae98ef10788da7a2b08aaf3bf660281bca/Source/SlimDetours/Trampoline.c#L40

The codebase also makes use of the lifted restriction in C99 that allows declaring variables anywhere, for example:
https://github.com/KNSoft/KNSoft.SlimDetours/blob/4cc236ae98ef10788da7a2b08aaf3bf660281bca/Source/SlimDetours/Transaction.c#L346-L349

The places where variables are declared in the beginning of the function are likely from times before C99. Since the codebase already requires C99, I think that new code should use these abilities, which make the code clearer and easier to reason about and maintain.

What do you think?

Reply:
Sorry I was mistaken, of course, follow the latest language standard is recommended.